src/OpenDefinitions.elm

```diff
@@ -1,165 +1,435 @@
+{-
+   OpenDefinitions is the main data structure for working with definitions in
+   the Workspace area. It supports holding Loading definitions and inserting
+   definitions after another.
+
+   It features a `focus` indicator and a before and after just like a Zipper,
+   but is based on OrderedDicts. `focus` can be changed with `next` and `prev`.
+
+   Invariants:
+     It structurally can't hold the invariant that it should not contain
+     duplicates, so this is instead enforced via the API; uniqueness is
+     determined by Hash.
+-}
+
+
 module OpenDefinitions exposing
-    ( OpenDefinitions
-    , definitions
+    ( HashIndexedDefinition
+    , OpenDefinitions
     , empty
-    , fromList
-    , get
+    , focus
+    , fromDefinitions
     , hashes
-    , insert
-    , insertAfter
-    , loading
+    , init
+    , insertWithFocus
+    , insertWithFocusAfter
+    , isEmpty
+    , isFocused
+    , map
+    , mapToList
     , member
+    , next
+    , prev
     , remove
     , replace
-    , replaceItems
+    , singleton
     , toList
-    , update
     )
 
 import Definition exposing (Definition)
 import Hash exposing (Hash)
+import List
+import List.Extra as ListE
 import OrderedDict exposing (OrderedDict)
 import RemoteData exposing (WebData)
 
 
+type alias DefinitionDict =
+    OrderedDict String (WebData Definition)
 
-{-
-   OpenDefinitions is the main data structure for working with definitions in
-   the Workspace area. It wraps an OrderedDirect (Hash to Definition), but is
-   opaque to that underlying data structure and only exposes a few key APIs,
-   including a few convenience functions for working with, loading and opening
-   Definitions.
--}
 
+type alias HashIndexedDefinition =
+    { hash : Hash
+    , definition : WebData Definition
+    }
 
+
+{-| This technically allows multiple of the same definition across the 3 fields.
+This is conceptionally not allowed and is enforced by the helper functions.
+-}
 type OpenDefinitions
-    = OpenDefinitions (OrderedDict String (WebData Definition))
+    = Empty
+    | OpenDefinitions
+        { before : DefinitionDict
```

Comment:
Using a `List HashIndexedDefinition` might be a way simpler implementation. Not sure if the `OrderedDict` is worth it.

Reply:
I've talked my self into at least giving a list a try. Will report back.

Reply:
Updated to be list based.

Reply:
I'd also probably favor simplicity of implementation, since this list is going to be tiny compared to how fast computers are -- even linear searches in a list that's hundreds of elements long is going to be instantaneous.